File: code/backend/Odin/src/main/kotlin/pt/isel/odin/service/department/DepartmentService.kt
package pt.isel.odin.service.department

import jakarta.transaction.Transactional
import org.springframework.stereotype.Service
import pt.isel.odin.http.controllers.department.models.SaveDepartmentInputModel
import pt.isel.odin.http.controllers.department.models.UpdateDepartmentInputModel
import pt.isel.odin.model.Department
import pt.isel.odin.repository.DepartmentRepository
import pt.isel.odin.service.department.error.DeleteDepartmentError
import pt.isel.odin.service.department.error.GetDepartmentError
import pt.isel.odin.service.department.error.SaveUpdateDepartmentError
import pt.isel.odin.utils.failure
import pt.isel.odin.utils.success

/**
 * Service for Departments.
 */
@Service
class DepartmentService(private val departmentRepository: DepartmentRepository) {

    /**
     * Gets a department by its id.
     *
     * @param id the department id
     *
     * @return the [GetDepartmentResult] if found, [GetDepartmentError.NotFoundDepartment] otherwise
     */
    fun getById(id: Long): GetDepartmentResult =
        departmentRepository.findById(id)
            .map<GetDepartmentResult> { depart -> success(depart) }
            .orElse(failure(GetDepartmentError.NotFoundDepartment))

    /**
     * Gets all departments.
     *
     * @return the [GetAllDepartmentsResult] with the list of [Department]
     */
    fun getAll(): GetAllDepartmentsResult = success(departmentRepository.findAll())

    /**
     * Saves a department.
     *
     * @param saveDepartInputModel the department to save
     *
     * @return the [CreationDepartmentResult] if saved, [SaveUpdateDepartmentError] otherwise
     */
    @Transactional
    fun save(saveDepartInputModel: SaveDepartmentInputModel): CreationDepartmentResult {
        if (departmentRepository.findByName(saveDepartInputModel.name).isPresent) {
            return failure(SaveUpdateDepartmentError.AlreadyExistsDepartment)
        }

        if (saveDepartInputModel.name.isBlank()) {
            return failure(SaveUpdateDepartmentError.IncorrectNameDepartment)
        }

        return success(departmentRepository.save(saveDepartInputModel.toDepartment()))
    }

    /**
     * Updates a department.
     *
     * @param updateDepartInputModel the department to update
     *
     * @return the [CreationDepartmentResult] if updated, [SaveUpdateDepartmentError] otherwise
     */
    @Transactional
    fun update(updateDepartInputModel: UpdateDepartmentInputModel): CreationDepartmentResult {
        if (updateDepartInputModel.name.isBlank()) {
            return failure(SaveUpdateDepartmentError.IncorrectNameDepartment)
        }

        return departmentRepository.findById(updateDepartInputModel.id)
            .map<CreationDepartmentResult> { depart ->
                success(
                    departmentRepository.save(
                        depart.copy(
                            name = updateDepartInputModel.name
                        )
                    )
                )
            }.orElse(failure(SaveUpdateDepartmentError.NotFoundDepartment))
    }

    /**
     * Deletes a department by its id.
     *
     * @param id the department id
     *
     * @return the [DeleteDepartmentResult] if deleted, [DeleteDepartmentError] otherwise
     */
    @Transactional
    fun delete(id: Long): DeleteDepartmentResult =
        departmentRepository.findById(id)
            .map<DeleteDepartmentResult> { depart ->
                departmentRepository.delete(depart)
                success(depart)
            }.orElse(failure(DeleteDepartmentError.NotFoundDepartment))
}

File: code/frontend/Tyr/src/services/department/DepartmentService.ts
import {apiRequest} from '../../axios/apiRequest';
import {
    CreationDepartmentResult,
    DeleteDepartmentResult,
    GetAllDepartmentsResult,
    GetDepartmentResult
} from "./DeparmentResult";
import {DepartmentInputModel} from "./models/DepartmentInputModel";

/**
 * Service to manage departments
 */
export namespace DepartmentService {

    const basePath = '/departments';

    /**
     * Get department by id
     *
     * @param id the id of the department
     *
     * @returns the department
     */
    export async function getById(id: number): Promise<GetDepartmentResult> {
        return apiRequest('get', `${basePath}/${id}`);
    }

    /**
     * Get all departments
     *
     * @returns all departments
     */
    export async function getAll(): Promise<GetAllDepartmentsResult> {
        return apiRequest('get', basePath);
    }

    /**
     * Save a department
     *
     * @param departmentRequest the department to save
     *
     * @returns the result of the creation
     */
    export async function save(departmentRequest: DepartmentInputModel): Promise<CreationDepartmentResult> {
        return apiRequest('post', `${basePath}/save`, departmentRequest);
    }

    /**
     * Update a department
     *
     * @param departmentRequest the department to update
     *
     * @returns the result of the update
     */
    export async function update(departmentRequest: DepartmentInputModel): Promise<CreationDepartmentResult> {
        return apiRequest('put', `${basePath}/update`, departmentRequest);
    }

    /**
     * Delete a department by id
     *
     * @param id the id of the department
     *
     * @returns the result of the deletion
     */
    export async function deleteById(id: number): Promise<DeleteDepartmentResult> {
        return apiRequest('delete', `${basePath}/${id}`);
    }
}

